packages/@uppy/companion/src/server/emitter/redis-emitter.js

```diff
@@ -20,54 +20,68 @@ function replacer(key, value) {
 module.exports = (redisClient, redisPubSubScope) => {
   const prefix = redisPubSubScope ? `${redisPubSubScope}:` : ''
   const getPrefixedEventName = (eventName) => `${prefix}${eventName}`
-  const publisher = redisClient.duplicate({ lazyConnect: true })
-  publisher.on('error', err => logger.error('publisher redis error', err.toString()))
-  /** @type {import('ioredis').Redis} */
-  let subscriber
-
-  const connectedPromise = publisher.connect().then(() => {
-    subscriber = publisher.duplicate()
-    subscriber.on('error', err => logger.error('subscriber redis error', err.toString()))
-    return subscriber.connect()
-  })
-
-  const handlersByEvent = new Map()
 
   const errorEmitter = new EventEmitter()
   const handleError = (err) => errorEmitter.emit('error', err)
 
-  connectedPromise.catch((err) => handleError(err))
+  async function makeRedis() {
+    const publisher = redisClient.duplicate({ lazyConnect: true })
+    publisher.on('error', err => logger.error('publisher redis error', err.toString()))
+    const subscriber = publisher.duplicate()
+    subscriber.on('error', err => logger.error('subscriber redis error', err.toString()))
+    await publisher.connect()
+    await subscriber.connect()
+    return { subscriber, publisher }
+  }
+
+  const redisPromise = makeRedis()
+  redisPromise.catch((err) => handleError(err))
 
+  /**
+   * 
+   * @param {(a: Awaited<typeof redisPromise>) => void} fn 
+   */
   async function runWhenConnected (fn) {
     try {
-      await connectedPromise
-      await fn()
+      await fn(await redisPromise)
     } catch (err) {
       handleError(err)
     }
   }
 
+  /** @type {Map<string, Map<() => unknown, () => unknown>>} */
+  const handlersByEvent = new Map()
+
   /**
    * Remove an event listener
    *
    * @param {string} eventName name of the event
    * @param {any} handler the handler of the event to remove
    */
-  function removeListener (eventName, handler) {
-    if (eventName === 'error') return errorEmitter.removeListener('error', handler)
+  async function removeListener (eventName, handler) {
+    if (eventName === 'error') {
+      errorEmitter.removeListener('error', handler)
+      return
+    }
 
-    return runWhenConnected(() => {
-      const handlersByThisEventName = handlersByEvent.get(eventName)
-      if (handlersByThisEventName == null) return undefined
+    const thisEventNameActualHandlerByHandler = handlersByEvent.get(eventName)
```

Comment:
actual handler by handler? 😅

Reply:
yes, it's a map that maps handler -> actual handler, meaning you can get the *actual handler* by providing *a handler*.

or in TS: `Map<() => void, () => void>`

where the key is the handler and the value is the actualHandler.

This naming was there before and i didn't have any better suggestion

Reply:
Can you write some docs above the `Map` why it is needed and why it has this structure?

Reply:
thinking about it more, this feels off to me? Why do we need handler -> handler? Is it because you have multiple handlers per event type? Maybe name it subscribers then and array of handlers per event name.

Reply:
because the api allows you to removeListener for a specific handler, then we need to keep a reference to that handler. i believe the redis-emitter's api mimics the EventEmitter API, and that api allows removing a specific handler from an event name. That's why it's implemented like this. We could probably change the API but that's a bigger change and potentially breaking, so maybe that could be done in a separate PR?

Reply:
added a comment and renamed the variables a bit